text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+              description: string_dictionary | string
+              unit: string_dictionary | string 
+              shared_attributes: *attributes                # attributes inherited by data points if not defined locally 
+              shared_start_time_unix_nano: timestamp        # start time inherited by data points if not defined locally
+              shared_time_unix_nano: timestamp              # required if not defined in data points
+              data:                                         # arrow type = sparse union
+                gauge:                                      # arrow type = struct 
+                    data_points: 
+                      - attributes: *attributes
+                        start_time_unix_nano: timestamp     # arrow type = timestamp (time unit nanoseconds)
+                        time_unix_nano: timestamp           # required if not defined as a shared field in the metric
+                        value:                              # arrow type = sparse union
+                          i64: int64 
+                          f64: float64 
+                        exemplars: *exemplars
+                        flags: uint32                       # each flag defined in this enum is a bit-mask
+                sum:                                        # arrow type = struct
+                    data_points: 
+                      - attributes: *attributes
+                        start_time_unix_nano: timestamp   
+                        time_unix_nano: timestamp           # required
+                        value:                              # arrow type = sparse union
+                          i64: int64
+                          f64: float64
+                        exemplars: *exemplars
+                        flags: uint32                         # each flag defined in this enum is a bit-mask
+                    aggregation_temporality: uint8_dictionary # OTLP enum with 3 variants
+                    is_monotonic: bool
+                summary:                                    # arrow type = struct
+                    data_points: 
+                      - attributes: *attributes
+                        start_time_unix_nano: timestamp
+                        time_unix_nano: timestamp           # required
+                        count: uint64
+                        sum: float64
+                        quantile:                           # arrow type = list of struct
+                          - quantile: float64
+                            value: float64
+                        flags: uint32                       # each flag defined in this enum is a bit-mask
+                histogram:                                  # arrow type = struct
+                    data_points:
+                      - attributes: *attributes
+                        start_time_unix_nano: timestamp
+                        time_unix_nano: timestamp
```

Comment:
I need to keep the semantic of the original OTLP protocol. Arrow duration data type is, if I'm not mistaken, represented as a single integer which will not allow me to find the two initial timestamps. The following text is an excerpt from the official OTEL documentation for these two fields to give you some additional context. 

// Both TimeUnixNano and StartTimeUnixNano values are expressed as
// UNIX Epoch time in nanoseconds since 00:00:00 UTC on 1 January 1970.
//
// # TimeUnixNano
//
// This field is required, having consistent interpretation across
// DataPoint types.  TimeUnixNano is the moment corresponding to when
// the data point's aggregate value was captured.
//
// Data points with the 0 value for TimeUnixNano SHOULD be rejected
// by consumers.
//
// # StartTimeUnixNano
//
// StartTimeUnixNano in general allows detecting when a sequence of
// observations is unbroken.  This field indicates to consumers the
// start time for points with cumulative and delta
// AggregationTemporality, and it should be included whenever possible
// to support correct rate calculation.  Although it may be omitted
// when the start time is truly unknown, setting StartTimeUnixNano is
// strongly encouraged.

Reply:
Gotcha, that makes sense.

> Arrow duration data type is, if I'm not mistaken, represented as a single integer which will not allow me to find the two initial timestamps.

Correct, you'd need a duration + one timestamp to find the other one, etc.

Reply:
I suppose it is possible (and would benefit compression) to encode one timestamp and a duration. This is a reversible translation, so it could be done for Arrow and un-done for standard OTLP.